File: bots/ska/lib/runtime-config.js
'use strict';

const fs = require('fs');
const path = require('path');

const DEFAULT_RUNTIME_CONFIG = {
  forecast: {
    conditionAdjustmentWeight: 0.50,
    reservationAdjustmentWeight: 0.42,
    shadowModelEnabled: true,
    shadowModelName: 'knn-shadow-v1',
    shadowNeighborCount: 7,
    shadowMinimumTrainRows: 21,
    shadowPromotionMapeGap: 2.0,
    sarimaPeriods: 7,
    sarimaMaxIter: 200,
    perModelAccuracyDays: 30,
    minimumModelWeight: 0.1,
    llmDiagnosisRagThreshold: 0.6,
    monthlyReviewGradeGood: 12,
    monthlyReviewGradeWarn: 22,
    weekdayBiasAlertAmount: 20000,
  },
  rebecca: {
    weeklyGradeGood: 10,
    weeklyGradeWarn: 20,
    anomalyRagThreshold: 0.55,
  },
  reviews: {
    daily: {
      minDays: 7,
      defaultDays: 30,
      avgMapeWarn: 20,
      avgMapeNotice: 12,
      avgBiasWarn: 30000,
      hitRate20Warn: 70,
      avgReservationGapWarn: 5,
      confidenceWarn: 0.4,
      upcomingDays: 3,
    },
    weekly: {
      minDays: 14,
      defaultDays: 56,
      avgMapeWarn: 20,
      avgMapeNotice: 12,
      avgBiasWarn: 150000,
      confidenceWarn: 0.4,
      weekdayMapeWarn: 20,
      upcomingDays: 7,
    },
  },
};

let cachedConfig = null;

function isObject(value) {
  return value && typeof value === 'object' && !Array.isArray(value);
}

function deepMerge(base, override) {
  if (!isObject(base) || !isObject(override)) return override ?? base;
  const merged = { ...base };
  for (const [key, value] of Object.entries(override)) {
    merged[key] = isObject(value) && isObject(base[key])
      ? deepMerge(base[key], value)
      : value;
  }
  return merged;
}

function loadSkaRuntimeConfig() {
  if (cachedConfig) return cachedConfig;
  try {
    const raw = JSON.parse(fs.readFileSync(path.join(__dirname, '..', 'config.json'), 'utf8'));
    cachedConfig = deepMerge(DEFAULT_RUNTIME_CONFIG, raw.runtime_config || {});
    return cachedConfig;
  } catch {
    cachedConfig = { ...DEFAULT_RUNTIME_CONFIG };
    return cachedConfig;
  }
}

function getSkaForecastConfig() {
  return loadSkaRuntimeConfig().forecast;
}

function getSkaRebeccaConfig() {
  return loadSkaRuntimeConfig().rebecca;
}

function getSkaReviewConfig() {
  return loadSkaRuntimeConfig().reviews;
}

module.exports = {
  loadSkaRuntimeConfig,
  getSkaForecastConfig,
  getSkaRebeccaConfig,
  getSkaReviewConfig,
};

File: bots/ska/src/runtime_config.py
import json
import os
from copy import deepcopy

DEFAULT_RUNTIME_CONFIG = {
    "forecast": {
        "conditionAdjustmentWeight": 0.50,
        "reservationAdjustmentWeight": 0.55,
        "calibrationMaxRatio": 0.22,
        "bookedHoursAdjustmentWeight": 0.40,
        "roomSpreadAdjustmentWeight": 0.24,
        "peakOverlapAdjustmentWeight": 0.22,
        "morningPatternAdjustmentWeight": 0.08,
        "afternoonPatternAdjustmentWeight": 0.12,
        "eveningPatternAdjustmentWeight": 0.18,
        "reservationTrendAdjustmentWeight": 0.24,
        "bookedHoursTrendAdjustmentWeight": 0.22,
        "shadowModelEnabled": True,
        "shadowModelName": "knn-shadow-v1",
        "shadowNeighborCount": 7,
        "shadowMinimumTrainRows": 21,
        "shadowPromotionMapeGap": 2.0,
        "sarimaPeriods": 7,
        "sarimaMaxIter": 200,
        "perModelAccuracyDays": 30,
        "minimumModelWeight": 0.10,
        "llmDiagnosisRagThreshold": 0.60,
        "monthlyReviewGradeGood": 12,
        "monthlyReviewGradeWarn": 22,
        "weekdayBiasAlertAmount": 20000,
    },
    "rebecca": {
        "weeklyGradeGood": 10,
        "weeklyGradeWarn": 20,
        "anomalyRagThreshold": 0.55,
    },
}

_cached_config = None


def _deep_merge(base, override):
    result = deepcopy(base)
    for key, value in (override or {}).items():
        if isinstance(value, dict) and isinstance(result.get(key), dict):
            result[key] = _deep_merge(result[key], value)
        else:
            result[key] = value
    return result


def load_runtime_config():
    global _cached_config
    if _cached_config is not None:
        return _cached_config

    config_path = os.path.join(os.path.dirname(__file__), '..', 'config.json')
    try:
        with open(config_path, 'r', encoding='utf-8') as fp:
            raw = json.load(fp)
        _cached_config = _deep_merge(DEFAULT_RUNTIME_CONFIG, raw.get('runtime_config') or {})
    except Exception:
        _cached_config = deepcopy(DEFAULT_RUNTIME_CONFIG)
    return _cached_config


def get_forecast_config():
    return load_runtime_config()["forecast"]


def get_rebecca_config():
    return load_runtime_config()["rebecca"]
